app/core/store/modules/levelSessions.js

```diff
@@ -0,0 +1,118 @@
+import levelSessionsApi from 'core/api/level-sessions'
+
+const SESSIONS_PER_REQUEST = 10
+
+export default {
+  namespaced: true,
+
+  state: {
+    loading: {
+      sessionsByClassroom: {}
+    },
+
+    // CLASSROOM_ID: {
+    //   sessions: [],
+    //   levelCompletionsByUser: {
+    //      LEVEL_ID: true / false
+    //    }
+    // }
+    levelSessionsByClassroom: {}
+  },
+
+  mutations: {
+    toggleClassroomLoading: (state, classroomId) => {
+      let loading = true;
+      if (state.loading.sessionsByClassroom[classroomId]) {
+        loading = false
+      }
+
+      Vue.set(state.loading.sessionsByClassroom, classroomId, loading)
+    },
+
+    initSessionsByClassroomState: (state, classroomId) => {
+      if (state.levelSessionsByClassroom[classroomId]) {
+        return
+      }
+
+      Vue.set(state.levelSessionsByClassroom, classroomId, {
+        sessions: [],
+        levelCompletionsByUser: {}
+      })
+    },
+
+    setSessionsForClassroom: (state, { classroomId, sessions }) => {
+      state.levelSessionsByClassroom[classroomId].sessions = sessions
+      state.levelSessionsByClassroom[classroomId].levelCompletionsByUser = {}
+    },
+
+    addLevelCompletionsByUserForClassroom: (state, { classroomId, levelCompletionsByUser }) => {
+      state.levelSessionsByClassroom[classroomId].levelCompletionsByUser = levelCompletionsByUser
+    }
+  },
+
+  // TODO add a way to clear out old level session data
+  actions: {
+    // TODO how do we handle two sets of parallel requests here? (ie user vists page, hits back and then visits again quickly)
```

Comment:
I think this is an argument for a Hot Cache on the server.
Each individual request promise could also have a `.then( // store in localStorage cache with a time)`.
And instead of just using `levelSessionsApi.fetchForClassroomMembers` there could be an initial check to the local cache for a recent result. (Maybe in last 2-5 minutes). Then the promise could resolve with this result if it exists.

I would be interested in any other ways this could be tackled.

Reply:
Agreed - adding a wrap up task to make sure we follow up on this https://app.asana.com/0/1115428071773315/1118364384628021/f